ultra/docs/curriculum_learning.md

```diff
@@ -0,0 +1,126 @@
+
+**Static curricula** in ULTRA currently use an implementation called the grade-based (GB) structure. The grade-based structure allows scenarios to be sequenced/ordered during training and testing runs. This is done by dividing an experiment into so called "grades", which are combinations of different tasks and levels. These grades are basically buckets for combinations of tasks and their levels.
+
+When an agent is put into training with curriculum-mode enabled (`ultra/train.py` is run with flag `--curriculum-mode True`), a coordinator object is initialized. It is responsible for setting up the scenarios in each grade and performing "graduation": switching of the agent into the next grade.
+
+Unlike the regular training setup (when curriculum mode is disabled), there is another task folder called `curriculum/` under `ultra/scenarios/` which has the blueprint of which set of scenarios are in each grade:
+
+```yaml
+curriculum:
+  static:
+    toggle: <bool> # Turn on static curriculum settings
+    grades:
+      1: [[<task>,<level>]]
+      2: [[<task>,<level>], [<task>,<level>]]
+      # 3: [[<task>,<level>], ... ,[<task>,<level>]] can have as many combinations of tasks and levels
+    conditions:
+      eval_per_grade: <bool> # Evaluation takes places after completion of grade. An "exam" at the end of year
+      episode_based: # Agent graduates after completing a N number of episodes in each grade
+        toggle: <bool> # Enable the condition
+        cycle: <bool> # Option to keep cycling through grades to episodes limit
+      pass_based: # Agent graduates after getting an average completion rate, the average is taken over the eval-rate (sampling-rate)
+        toggle: <bool> # Enable the condition
+        pass_rate: <float> # Scalar between 0 and 1; describes the threshold completion rate (%)
+        sample_rate: <int> # Samples for some parameter (average_reached_goal) at every N episodes
+        warmup_episodes: <int> # Starting N episodes where sampling does not occur, thus no graduation can take place
+```
+A more specific example in which we take the three levels from task 1 and distribute it among three grades:
+
+```yaml
+curriculum:
+  static:
+    toggle: True
+    grades:
+      1: [[1,no-traffic]]
+      2: [[1,easy]]
+      3: [[1,hard]]
+      conditions:
+        eval_per_grade: True
+        episode_based:
+          toggle: True 
+          cycle: True 
+        pass_based:
+          toggle: False 
+          pass_rate: 0.50 # If the average scenario passed exceeds more than 0.5 then grade is switched
+          sample_rate: 30 # Every 30 episodes slot, the average reached goal (arg) is calculated (arg = total_scenario_passed / sample_rate)
+          warmup_episodes: 100 # Starting at any grade, the first 100 episodes will be subjected to no sampling
```

Comment:
These are ignored if toggle = false, correct?

Reply:
Yes, these are ignored if the pass_based_toggle is true